Implements a booking request model

MilesDev91 opened the pull request:
Added the request model, complete with validation.

### What this PR does:
- Adds the Request model to the server side of the application.

### Any information needed to test this feature:
- Check the Request.js file in the models folder

### Review thoughts
- I read a bit more into validation, so please let me know if I missed something.
- I included an id field for the pets, but if we aren't handling pets in that way I can remove it.

closes #9

MilesDev91 committed:
Request model

Added the request model, complete with validation.

rajivtitus commented on server/models/Request.js:
```diff
@@ -0,0 +1,28 @@
+const requestSchema = new mongoose.Schema({
+  user_id: { type: String, max: 50, required: true },
+  sitter_id: { type: String, max: 50, required: true },
+  pet_ids: { type: [String], max: 50, required: true },
```
This is a comment

ugur-akin commented on server/models/Request.js:
```diff
@@ -0,0 +1,28 @@
+const mongoose = require("mongoose");
+
+const requestSchema = new mongoose.Schema({
+  user_id: { type: String, max: 50, required: true },
+  sitter_id: { type: String, max: 50, required: true },
+  pet_ids: { type: [String], max: 50, required: true },
```
Since these fields refer to another table/collection, it would be best to make them `ref`s instead (you can find more info [here](https://mongoosejs.com/docs/populate.html#populate)).

Pets are a nice addition 🎉

MilesDev91 replied:
Forgot about that, thanks

MilesDev91 replied:
Also, should I call it petNames instead of petIds or pets. Don't know if we are going to have any more information besides the name of the pet.

rajivtitus replied:
@MilesDev91 I think we call it `petIds` - at a later date we can add in other information like breed, size, age etc to the Pet model  😄

ugur-akin commented on server/models/Request.js:
```diff
@@ -0,0 +1,28 @@
+  user_id: { type: String, max: 50, required: true },
+  sitter_id: { type: String, max: 50, required: true },
+  pet_ids: { type: [String], max: 50, required: true },
+  start_date: {
+    type: Date,
+    validate: {
+		validator: function(startDate) { return startDate < this.end_date },
+		message: 'Start date must begin before the end date.'
+    },
+	required: [true, 'Must have a start date.']
+  },
+  end_date: {
+    type: Date,
+    validate: {
+		validator: function(endDate) { return this.start_date < endDate },
+		message: 'Start date must begin before the end date.'
+    },
+	required: [true, 'Must have an end date.']
+  },
```
Validations are always nice to see 💯

ugur-akin commented on server/models/Request.js:
```diff
@@ -0,0 +1,28 @@
+    },
+	required: [true, 'Must have an end date.']
+  },
+  accepted: { type: Boolean, default: False },
+  declined: { type: Boolean, default: False },
```
Having multiple fields that needs to be synchronized with each other usually leads to unnecessary complexity - can we simplify these into a single field (e.g. `status` with some [string `SchemaTypes`](https://mongoosejs.com/docs/schematypes.html#string-validators)) ?

MilesDev91 replied:
So, would a status field with a String enum with 'accepted' and 'declined' be appropriate?

rajivtitus replied:
Yes, that would be a good idea. We can have `['pending', 'accepted', 'declined', 'completed']` as the options for the status of a request. As for the `paid` field we can have it seperate from the status as we only want it to be accessed through the payment routes 👍🏽

ugur-akin commented on server/models/Request.js:
```diff
@@ -0,0 +1,28 @@
+  accepted: { type: Boolean, default: False },
+  declined: { type: Boolean, default: False },
+  paid: { type: Boolean, default: False }
+});
```
I'd also encourage adding [timestamps](https://mongoosejs.com/docs/guide.html#timestamps) to your schemas as it's a good habit to have 👍🏼

MilesDev91 replied:
Good thinking

ugur-akin reviewed (changes requested):
Good job Miles, this looks solid! 💯 I mentioned a few things that can be polished further, please have a look.

For mongoose field names, we generally follow the `camelCase` convention. Could we please change the fields here accordingly?

Lastly, clearly communicating the details of your pull requests is a priceless skill to have and I think we can do more to build the right habits on that front. I would recommend paying great attention to the pr title and the pr body alike, some key points would be:
- using proper English in the title (e.g. `Implements a model for booking requests`)
- [mention/link the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) in your pull request description (e.g. `closes #9`), you can also consider adding any labels you see fit to your pull request,
- formatting the pr desctiption cleanly, using proper English sentences, removing any unused sections and/or template details such as `(required)` ✅ 

I would definitely recommend checking out [this article](https://hugooodias.medium.com/the-anatomy-of-a-perfect-pull-request-567382bb6067) as it highlights important points about how best to format pull requests 👍🏼

MilesDev91 committed:
Update Request.js

Added a pet schema for the pets field.

Changed some of the field names to better clarify their purpose and contents. Also fixed the casing.

Added the timestamps option so we have a timestamp of creation and updates.

MilesDev91 commented:
I really appreciate the feedback! Hope the changes are moving in the right direction.

MilesDev91 committed:
Update Request.js

Just added a couple values for the status and changed a name

ugur-akin commented on server/models/Request.js:
```diff
@@ -1,28 +1,47 @@
     validate: {
-		validator: function(startDate) { return startDate < this.end_date },
-		message: 'Start date must begin before the end date.'
+      validator: function(startDate) { return startDate < this.end_date },
```
```suggestion
      validator: function(startDate) { return startDate < this.endDate },
```
👍🏼

ugur-akin commented on server/models/Request.js:
```diff
@@ -1,28 +1,47 @@
     validate: {
-		validator: function(endDate) { return this.start_date < endDate },
-		message: 'Start date must begin before the end date.'
+      validator: function(endDate) { return this.start_date < endDate },
```
```suggestion
      validator: function(endDate) { return this.startDate < endDate },
```

ugur-akin commented on server/models/Request.js:
```diff
@@ -1,28 +1,47 @@
   },
-  accepted: { type: Boolean, default: False },
-  declined: { type: Boolean, default: False },
+  // Tracks if a request has yet been accepted or declined
```
I think the field is pretty self-explanatory so I don't think this comment is really adding much. Can we remove it?

ugur-akin commented on server/models/Request.js:
```diff
@@ -1,28 +1,47 @@
-  accepted: { type: Boolean, default: False },
-  declined: { type: Boolean, default: False },
+  // Tracks if a request has yet been accepted or declined
+  status: { type: String, default: 'pending', enum: ['accepted', 'declined', 'pending', 'completed'] },
```
I'd also recommend using more [string options](https://mongoosejs.com/docs/schematypes.html#string-validators) here such as `trim/lowercase` in order to provide a more flexibile interface.

MilesDev91 replied:
Good idea for any string enum I suppose, didn't think about that.

ugur-akin reviewed (approved):
LGTM! Just a few minor issues to fix. I'll go ahead and approve this but please fix the highlighted typos and issues before merging 🙏🏼 

Thanks for your diligent work!

MilesDev91 committed:
Update Request.js

Made some finishing touches

MilesDev91 committed:
Merge branch 'main' into request_model

MilesDev91 committed:
Update Request.js
